inc/geometries/nablas.h

```diff
@@ -0,0 +1,114 @@
+#pragma once
+#include <functional>
+#include "dg/backend/memory.h"
+#include "dg/topology/geometry.h"
+
+#include "../src/feltor/feltor.h"
+#include "../src/feltor/parameters.h"
+ //#include "feltor/init.h"
+
+ namespace dg
+ {
+ namespace geo
+ {
+ ///@addtogroup fluxfunctions
+ /**
+  * @brief Certain functions related with using Nabla (in divergences, perpendicular gradients, vector dot nabla...)
+  * @ingroup misc_geo
+  */
+
+ template<class Geometry, class Container, class Matrix>                                                                                                                            
+ struct Nablas 
+ {
+    using geometry_type = Geometry;
+    using matrix_type = Matrix;
+    using container_type = Container;
+    using value_type = get_value_type<Container>;
+
+ /**
+  * @brief Main contructor: construct from a 3D geometry and also with a Magnetic field and a bool parameter in case there is reversed field
+  * @tparam Geometry
+  * @param geom3d
+  * @tparam g::geo::TokamakMagneticField
+  * @param mag
+  * @tparam  bool
+  * @param reversed
+  */
+
+ Nablas(const Geometry& geom3d, dg::geo::TokamakMagneticField& mag, bool reversed): m_g(geom3d), m_mag(mag), m_reversed_field(reversed) {
+     dg::blas2::transfer( dg::create::dx( m_g, dg::DIR, dg::centered), m_dR);
+     dg::blas2::transfer( dg::create::dy( m_g, dg::DIR, dg::centered), m_dZ);
+     m_vol=dg::tensor::volume(m_g.metric());
+     m_tmp=m_tmp2=m_tmp3=m_vol;
+     m_hh=m_g.metric();
+
+     auto bhat = dg::geo::createBHat( m_mag);
+     bhat = dg::geo::createEPhi(+1);
+     if( m_reversed_field)
+     bhat = dg::geo::createEPhi(-1);
+     m_hh = dg::geo::createProjectionTensor( bhat, m_g);
```

Comment:
Are you aware that without the toroidal field approx. this will make the actual perp tensor g - bb and thus the grad_perp_f and v_dot_nabla_f functions try to compute the actual perpendicular gradients? (in a wrong way because the third direction is missing)? Also I don't think that the reversed field parameter changes anything because g- bb is still the same whether b is negative or positive.

Reply:
If this becomes too tedious and confusing maybe remove the TokamakMagnetic field altogether and just compute with dR and dZ